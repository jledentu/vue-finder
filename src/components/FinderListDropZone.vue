<template>
  <div
    class="drop-zone"
    :class="{ 'drag-over': dragOver }"
    :style="{
      ...(dragOver &&
        theme.primaryColor && { borderColor: theme.primaryColor }),
      ...(dragOver &&
        theme.dropZoneBgColor && { backgroundColor: theme.dropZoneBgColor })
    }"
    @dragenter.prevent="onDragEnter"
    @dragleave.prevent="onDragLeave"
    @dragover.prevent
    @drop.prevent="onDrop"
  />
</template>

<script>
import { get } from "lodash-es";

export default {
  name: "FinderListDropZone",
  props: {
    treeModel: {
      type: Object,
      required: true
    },
    node: {
      type: Object,
      required: true
    },
    dragEnabled: {
      type: Boolean,
      default: false
    },
    options: {
      type: Object,
      default: () => ({})
    }
  },
  data: () => ({
    dragCounter: 0
  }),
  computed: {
    dragOver() {
      return this.dragCounter > 0;
    },
    theme() {
      return get(this, "options.theme", {});
    }
  },
  methods: {
    onDragEnter() {
      if (this.treeModel.isDragging()) {
        this.dragCounter++;
      }
    },
    onDragLeave() {
      if (this.treeModel.isDragging()) {
        this.dragCounter--;
      }
    },
    onDrop() {
      if (!this.treeModel.isDragging()) {
        return;
      }
      this.dragCounter = 0;
      this.treeModel.dropOnNode(this.node.id);
    }
  }
};
</script>

<style lang="scss" scoped>
@import "../style/colors";

.drop-zone {
  transition: height 0.1s ease;
  height: 0.5em;
  overflow: hidden;

  &.drag-over {
    opacity: 0.5;
    margin-bottom: 0;
    height: 2em;
    border: dashed 3px $primaryColor;
    background-color: change-color($primaryColor, $alpha: 0.2);
  }
}
</style>
