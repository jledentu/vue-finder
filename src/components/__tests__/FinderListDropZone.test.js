import { mount } from "@vue/test-utils";
import TreeModel from "@/utils/tree-model";
import FinderListDropZone from "../FinderListDropZone";

jest.mock("@/utils/tree-model");

describe("FinderListDropZone", () => {
  let treeModel;
  let node;
  const tree = {
    id: "test1",
    children: [
      {
        id: "test11",
        selected: true,
        children: [
          {
            id: "test111"
          },
          {
            id: "test112"
          }
        ]
      },
      {
        id: "test12"
      }
    ]
  };
  beforeEach(() => {
    treeModel = new TreeModel(tree);
    node = {
      id: "test111"
    };
  });

  it("should match snapshot", () => {
    const wrapper = mount(FinderListDropZone, {
      propsData: {
        treeModel,
        node
      }
    });
    expect(wrapper).toMatchSnapshot();
  });

  it("should match if drag enter and is dragging", () => {
    treeModel.isDragging.mockReturnValue(true);
    const wrapper = mount(FinderListDropZone, {
      propsData: {
        treeModel,
        node,
        dragEnabled: true
      }
    });
    wrapper.trigger("dragenter");
    expect(wrapper).toMatchSnapshot();
  });

  it("should match if drag leave and is dragging", () => {
    treeModel.isDragging.mockReturnValue(true);
    const wrapper = mount(FinderListDropZone, {
      propsData: {
        treeModel,
        node,
        dragEnabled: true
      }
    });
    wrapper.trigger("dragenter");
    wrapper.trigger("dragleave");
    expect(wrapper).toMatchSnapshot();
  });

  it("should match if drag enter and not dragging", () => {
    treeModel.isDragging.mockReturnValue(false);
    const wrapper = mount(FinderListDropZone, {
      propsData: {
        treeModel,
        node
      }
    });
    wrapper.trigger("dragenter");
    expect(wrapper).toMatchSnapshot();
  });

  it("should match if drag leave and not dragging", () => {
    treeModel.isDragging.mockReturnValue(false);
    const wrapper = mount(FinderListDropZone, {
      propsData: {
        treeModel,
        node
      }
    });
    wrapper.trigger("dragenter");
    wrapper.trigger("dragleave");
    expect(wrapper).toMatchSnapshot();
  });

  describe("#onDrop", () => {
    it("should call `treeModel.dropOnNode`", () => {
      treeModel.isDragging.mockReturnValue(true);
      const wrapper = mount(FinderListDropZone, {
        propsData: {
          treeModel,
          node,
          dragEnabled: true
        }
      });
      wrapper.trigger("drop");
      expect(treeModel.dropOnNode).toHaveBeenCalledWith("test111");
    });

    it("should not call `treeModel.dropOnNode` if not dragging", () => {
      treeModel.isDragging.mockReturnValue(false);
      const wrapper = mount(FinderListDropZone, {
        propsData: {
          treeModel,
          node,
          dragEnabled: true
        }
      });
      wrapper.trigger("drop");
      expect(treeModel.dropOnNode).not.toHaveBeenCalled();
    });
  });
});
