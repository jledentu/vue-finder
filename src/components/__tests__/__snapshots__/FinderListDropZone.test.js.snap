// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FinderListDropZone should match if drag enter and is dragging 1`] = `<div class="drop-zone drag-over"></div>`;

exports[`FinderListDropZone should match if drag enter and not dragging 1`] = `<div class="drop-zone"></div>`;

exports[`FinderListDropZone should match if drag leave and is dragging 1`] = `<div class="drop-zone"></div>`;

exports[`FinderListDropZone should match if drag leave and not dragging 1`] = `<div class="drop-zone"></div>`;

exports[`FinderListDropZone should match snapshot 1`] = `<div class="drop-zone"></div>`;
